### PR TITLE
ADBDEV-3209: Add possibility to decrypt an encrypted password

### PR DESCRIPTION
**Pull request by RomaZe**

### CODE REVIEW

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package org.greenplum.pxf.api.configuration;
+
+import io.arenadata.security.encryption.client.configuration.JksTextEncryptorConfiguration;
+import org.springframework.beans.factory.annotation.Value;
+
```
Comment (iamlapa) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java:7]:
```java
@Configuration
@ConditionalOnProperty({"pxf.ssl.jks-store.path", "pxf.ssl.jks-store.password", "pxf.ssl.salt.key"})
```

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class PxfJksTextEncryptorConfiguration extends JksTextEncryptorConfiguration {
+    @Value("${pxf.ssl.jks-store.path}")
```
Comment (iamlapa) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: use this annotation in constructor

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    private String path;
+
+    @Value("${pxf.ssl.jks-store.password}")
```
Comment (iamlapa) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: same

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    private String password;
+
+    @Value("${pxf.ssl.salt.key}")
```
Comment (iamlapa) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: same

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 2 above)
Comment (iamlapa) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: private final ...

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java [hunk 5]
```diff
@@ -335,11 +337,24 @@ public void afterPropertiesSet() {
             );
         }
 
-        // This must be the last parameter parsed, as we output connectionConfiguration earlier
-        // Optional parameter. By default, corresponding connectionConfiguration property is not set
         if (jdbcUser != null) {
             String jdbcPassword = configuration.get(JDBC_PASSWORD_PROPERTY_NAME);
             if (jdbcPassword != null) {
+                if (jdbcPassword.startsWith(Util.getEncryptedMessagePrefix(EncryptorType.AES256))) {
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java]: 1. Add a `DecryptClient` field
2. Get a `DecryptClient` from `SpringContext` in constructor
3. Use a `DecryptClient` here: `jdbcPassword = DecryptClient == null? jdbcPassword : DecryptClient.decrypt(jdbcPassword)`

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/JdbcDecryptService.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package org.greenplum.pxf.plugins.jdbc.utils;
+
+import io.arenadata.security.encryption.client.provider.TextEncryptorProvider;
+import io.arenadata.security.encryption.client.service.DecryptClient;
+import org.greenplum.pxf.api.configuration.PxfJksTextEncryptorConfiguration;
+
+public class JdbcDecryptService {
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/JdbcDecryptService.java]: Unnecessary class

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/spring/PxfJksTextEncryptor.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package org.greenplum.pxf.service.spring;
+
+import org.greenplum.pxf.api.configuration.PxfJksTextEncryptorConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
```
Comment (iamlapa) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/spring/PxfJksTextEncryptor.java]: Unnecessary class

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 1 above)
Comment (RomaZe) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java:7]: Added annotation to this class. 

But it seems that if we have wrong parameter values or there is some error to get encryption key from the store we will receive error to load bean and PXF service will not start. Is it expected behavior?

`org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'textEncryptorProvider' defined in class path resource [org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.class]`

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 2 above)
Comment (RomaZe) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: Done

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 3 above)
Comment (RomaZe) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: Done

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 4 above)
Comment (RomaZe) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java]: Done

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java (hunk 5 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java]: Done, but I don't fully understand why we should use `DecryptClient` as a field.
Why we cannot use it just as a local variable?

One more concern. If we don't check whether the password is encrypted or not in the pxf-jdbc file we will always decrypt password for any jdbc connection in different profiles while `DecryptClient` is not null. So we will not have an opportunity to select whether to save password as encrypted or not. We will have to select only one option for all jdbc passwords.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/JdbcDecryptService.java (hunk 6 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/JdbcDecryptService.java]: removed

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/spring/PxfJksTextEncryptor.java (hunk 7 above)
Comment (RomaZe) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/spring/PxfJksTextEncryptor.java]: removed

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java (hunk 1 above)
Comment (iamlapa) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/configuration/PxfJksTextEncryptorConfiguration.java:7]: Yes, of course! We should throw an exception as early as possible if user has entered invalid settings.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java (hunk 5 above)
Comment (iamlapa) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java]: 1. Why we cannot use it just as a local variable? - SOLID -> S -> Single Responsibility Principle -> If we refuse a class field in favor of a local variable(from the static method `org.greenplum.pxf.api.utilities.SpringContext#getBean`), then we implicitly shift the responsibility for password decryption to `JdbcBasePlugin`, thereby violating the principle. We also make it harder to test this class 
2. The plugin should not check whether the password is encrypted or not, let the `DecryptClient` be responsible for this

Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcBasePlugin.java]: Thank you for clarification!

File: server/build.gradle [hunk 8]
```diff
@@ -114,6 +138,9 @@ configure(javaProjects) {
             dependency("org.wildfly.openssl:wildfly-openssl:1.0.7.Final")
             dependency("org.xerial.snappy:snappy-java:1.1.8.4")
 
+            // Arenadata encryption
+            dependency("io.arenadata.security:encryption:1.0.0-SNAPSHOT")
```
Comment (iamlapa) [inline server/build.gradle]: remove "snapshot"

Comment (RomaZe) [reply server/build.gradle]: done

**Review by iamlapa — APPROVED** [review]

**Pull request merged**